contracts/chainlink/OracleHubChainLink.sol

```diff
@@ -0,0 +1,142 @@
+
+    /// @dev called by our factory to create the needed initial query for a new MarketContract
+    /// @param marketContractAddress - address of the newly created MarketContract
+    /// @param oracleQueryURL   URL of rest end point for data IE 'https://api.kraken.com/0/public/Ticker?pair=ETHUSD'
+    /// @param oracleQueryPath  path of data inside json object. IE 'result.XETHZUSD.c.0'
+    function requestQuery(
+        address marketContractAddress,
+        string oracleQueryURL,
+        string oracleQueryPath
+    ) external onlyFactory
+    {
+        ChainLinkQuery storage chainLinkQuery = contractAddressToChainLinkQuery[marketContractAddress];
+        chainLinkQuery.marketContractAddress = marketContractAddress;
+        chainLinkQuery.oracleQueryURL = oracleQueryURL;
+        chainLinkQuery.oracleQueryPath = oracleQueryPath;
+
+        // NOTE: currently this cannot be separated into its own function due to constraints around
+        // return types and the current version of solidity... is this true for internal / private fxs?
+        StringLib.slice memory pathSlice = oracleQueryPath.toSlice();
+        StringLib.slice memory delim = ".".toSlice();
+        chainLinkQuery.oracleQueryRunPath = new string[](pathSlice.count(delim) + 1);
+        for (uint i = 0; i < chainLinkQuery.oracleQueryRunPath.length; i++) {
+            chainLinkQuery.oracleQueryRunPath[i] = pathSlice.split(delim).toString();
+        }
+        bytes32 jobId; // TODO: how do we want to generate these?
+        createRunAndRequest(jobId, chainLinkQuery);
+    }
+
+    /// @dev designated callback function to ChainLinks oracles.
+    /// @param requestId id of the request being handed back
+    /// @param price     value being handed back from oracle.
+    function callback(bytes32 requestId, uint256 price) external checkChainlinkFulfillment(requestId) {
+        // NOTE: event is emitted by chainlink upon call above to checkChainlinkFulfillment
+        ChainLinkQuery memory chainLinkQuery = requestIDToChainLinkQuery[requestId];
+        require(chainLinkQuery.marketContractAddress != address(0), "market contract address can not be null!");
+        MarketContractChainLink(chainLinkQuery.marketContractAddress).oracleCallBack(price);
+    }
+
+    /*
+    // PRIVATE METHODS
+    */
+
+    /// @dev Creates the needed Run object and initiates the request to ChainLink.
+    function createRunAndRequest(bytes32 jobId, ChainLinkQuery memory chainLinkQuery) private returns (bytes32) {
+        ChainlinkLib.Run memory run = newRun(jobId, this, "callback(bytes32,uint256)");
+        run.add("url", chainLinkQuery.oracleQueryURL);
+        run.addStringArray("path", chainLinkQuery.oracleQueryRunPath);
+        run.addUint("times", MarketContract(chainLinkQuery.marketContractAddress).PRICE_DECIMAL_PLACES());
+        bytes32 requestID = chainlinkRequest(run, LINK(1));
+        // TODO: figure out handling of JOB_ID and REQUEST_ID
+        // TODO: add sleep adapter
```

Comment:
Run parameters for the sleep adapter can be added like so:
```
run.addUint("until", now + 1 hours);
```
You'll just need to use `run.addUint` with the key `"until"` and a Unix timestamp as the value.

Also, a job which uses the sleep adapter would have a different JobID than one which doesn't. You will still be able to pass in the "until" key as a run parameter without harming anything for the job without the sleep adapter, however, it'll be costing you gas for essentially no benefit.

Reply:
:+1: